sched/scheduler/cluster_state.go

```diff
@@ -103,26 +129,64 @@ func (c *clusterState) updateCluster() {
 	}
 }
 
-// Processes nodes being added and removed from the cluster &
-// updates the distributor state accordingly
+// Processes nodes being added and removed from the cluster & updates the distributor state accordingly.
+// Note, we don't expect there to be many updates after startup if the cluster is relatively stable.
 func (c *clusterState) update(updates []cluster.NodeUpdate) {
+	// Apply updates
 	for _, update := range updates {
 		switch update.UpdateType {
 		case cluster.NodeAdded:
-			// add the node if it doesn't already exist
-			if _, ok := c.nodes[update.Node.Id()]; !ok {
-				c.nodes[update.Node.Id()] = newNodeState(update.Node)
-				c.nodeGroups[""].idle[update.Node.Id()] = update.Node
-				log.Infof("Added node: %+v, now have %d nodes\n", update.Node, len(c.nodes))
+			if ns, ok := c.suspendedNodes[update.Id]; ok {
+				// This node was suspended earlier (assuming id is unique to a running node instance), we can recover it now.
+				ns.timeLost = nilTime
+				c.nodes[update.Id] = ns
+				delete(c.suspendedNodes, update.Id)
+				log.Infof("Recovered suspended node %v (%v), now have %d healthy nodes, %d suspended nodes",
+					update.Id, ns, len(c.nodes), len(c.suspendedNodes))
+			} else if ns, ok := c.nodes[update.Id]; !ok {
+				// This is a new unrecognized node, add it to the cluster.
+				c.nodes[update.Id] = newNodeState(update.Node)
+				c.nodeGroups[""].idle[update.Id] = c.nodes[update.Id]
+				log.Infof("Added new node: %v (%+v), now have %d nodes", update.Id, update.Node, len(c.nodes))
+			} else {
+				// This node is already present, log this spurious add.
```

Comment:
is this a symptom of a bigger issue?  Should we log it as an 'ignored' error - use Error level, but indicate that this error was ignored.  Same with ln 170 below.